graspy/match/sgm.py

```diff
@@ -0,0 +1,332 @@
+# Copyright 2019 NeuroData (http://neurodata.io)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import math
+from scipy.optimize import linear_sum_assignment
+from scipy.optimize import minimize_scalar
+from sklearn.utils import check_array
+from sklearn.utils import column_or_1d
+from .skp import SinkhornKnopp
+
+
+class SeededGraphMatching:
+    """
+    Seeded Graph Matching Algorithm (SGM)
+    The seeded graph matching problem is a variation of graph matching
+     in which part of the matching is fixed. This algorithm is a modification
+     of FAQ, an algorithm also implemented in this package[1].
+
+
+    Parameters
+    ----------
+
+    n_init : int, positive (default = 1)
+        Number of random initializations of the starting permutation matrix that
+        the FAQ algorithm will undergo. n_init automatically set to 1 if
+        init_method = 'barycenter'
+
+    init_method : string (default = 'barycenter')
+        The initial position chosen
+
+        "barycenter" : the non-informative “flat doubly stochastic matrix,”
+        :math:`J=1*1^T /n` , i.e the barycenter of the feasible region
+
+        "rand" : some random point near :math:`J, (J+K)/2`, where K is some random doubly
+        stochastic matrix
+
+    max_iter : int, positive (default = 30)
+        Integer specifying the max number of FW iterations.
+        FAQ typically converges with modest number of iterations
+
+    shuffle_input : bool (default = True)
+        Gives users the option to shuffle the nodes of A matrix to avoid results
+        from inputs that were already matched
+
+    eps : float (default = 0.1)
+        A positive, threshold stopping criteria such that FW continues to iterate
+        while Frobenius norm of :math:`(P_{i}-P_{i+1}) > eps`
+
+
+    gmp : bool (default = False)
+        Gives users the option to the Graph Matching Problem (GMP) rather than
+        the Quadratic Assignment (QAP). This is accomplished through trivial
+        negation of the objective function.
+
+    Attributes
+    ----------
+
+    perm_inds_ : array, size (n,) where n is the number of vertices in the graphs fitted.
+        The indices of the optimal permutation on the nodes of B, found via
+        FAQ, to best minimize the objective function :math:`f(P) = trace(A^T PBP^T )`.
```

Comment:
specify that permutation inds respect seeds as well (i.e. those nodes won't be permuted)